bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/scheduler2/InternalSchedulerImpl.java

```diff
@@ -0,0 +1,320 @@
+        ((ThreadPoolExecutor) executor).allowCoreThreadTimeOut(true);
+    }
+
+    @Deactivate
+    void deactivate() {
+        List<Runnable> shutdownNow = executor.shutdownNow();
+        if (shutdownNow != null && shutdownNow.size() > 0) {
+            logger.warn("Shutdown executables {}", shutdownNow);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Instant> after(Duration duration) {
+        CompletableFuture<Instant> deferred = new CompletableFuture<>();
+        Instant start = Instant.now();
+        ScheduledFuture<?> schedule = executor.schedule(() -> {
+            deferred.complete(start);
+        }, duration.toMillis(), TimeUnit.MILLISECONDS);
+        deferred.exceptionally((Throwable e) -> {
+            if (e instanceof CancellationException) {
+                schedule.cancel(true);
+            }
+            return null;
+        });
+        return deferred;
+    }
+
+    @Override
+    public <T> CompletableFuture<T> after(Callable<T> callable, Duration duration) {
+        CompletableFuture<T> deferred = new CompletableFuture<>();
+        ScheduledFuture<?> schedule = executor.schedule(() -> {
+            try {
+                deferred.complete(callable.call());
+            } catch (Throwable e) {
+                deferred.completeExceptionally(e);
+            }
+        }, duration.toMillis(), TimeUnit.MILLISECONDS);
+        deferred.exceptionally((Throwable e) -> {
+            if (e instanceof CancellationException) {
+                schedule.cancel(true);
+            }
+            return null;
+        });
+        return deferred;
+    }
+
+    static class Unique {
+        AtomicBoolean done = new AtomicBoolean();
+
+        interface RunnableException {
+            public void run();
+        }
+
+        boolean once(RunnableException o) {
+            if (done.getAndSet(true) == false) {
+                o.run();
+                return true;
+            } else {
+                return false;
+            }
+        }
+    }
+
+    @Override
+    public <T> CompletableFuture<T> before(CompletableFuture<T> promise, Duration timeout) {
+        CompletableFuture<T> d = new CompletableFuture<T>();
+        Unique only = new Unique();
+
+        after(timeout).thenAccept((p) -> {
+            only.once(() -> d.completeExceptionally(TimeoutException.SINGLETON));
+        });
+
+        promise.thenAccept((p) -> {
+            only.once(() -> d.complete(p));
+        }).exceptionally((Throwable ex) -> {
+            only.once(() -> d.completeExceptionally(ex));
+            return null;
+        });
+        return d;
+    }
+
+    private static abstract class Schedule {
+        volatile CompletableFuture<?> promise;
+        volatile boolean canceled;
+        Throwable exception;
+
+        abstract Instant next();
+
+        abstract void doIt() throws Exception;
+    }
+
+    private static class PeriodSchedule extends Schedule {
```

Comment:
If we want to introduce other types of expressions, one might consider to remove this class altogether, and rather craft a TemporalAdjuster that deals with the desired Period instead.

Reply:
Good point! That was one of the open things to clean up which I wanted to do only if this way gets more or less the "approval" of everybody. I quickly did it nevertheless: It indeed turned out nicely, I think.